File: libs/power-log-parser/src/lib/parsers/mercenaries-ability-activated-parser.ts
import { BlockType, CardType, GameTag, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Action, Node, NodeType } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class MercenariesAbilityActivatedParser implements ActionParser {
	readonly ParserName = 'MercenariesAbilityActivatedParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		let action: Action | null = null;
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.Action &&
			(action = node.Object as Action).Type === (BlockType.PLAY as number) &&
			this.GameState.CurrentEntities.has(action.Entity) &&
			this.GameState.CurrentEntities.get(action.Entity)!.GetZone() === (Zone.LETTUCE_ABILITY as number) &&
			this.GameState.CurrentEntities.get(action.Entity)!.GetCardType() === (CardType.LETTUCE_ABILITY as number)
		);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const action = node.Object as Action;
		const entity = this.GameState.CurrentEntities.get(action.Entity)!;
		const controllerId = entity.GetEffectiveController();
		const cardId = entity.CardId;
		const abilityOwnerEntityId = entity.GetTag(GameTag.LETTUCE_ABILITY_OWNER);
		const isTreasure = entity.GetTag(GameTag.LETTUCE_IS_TREASURE_CARD) === 1;
		return [
			GameEventProvider.Create(
				action.TimeStamp,
				'MERCENARIES_ABILITY_ACTIVATED',
				GameEventHelper.CreateProvider(
					'MERCENARIES_ABILITY_ACTIVATED',
					cardId,
					controllerId,
					entity.Id,
					this.StateFacade,
					{
						AbilityOwnerEntityId: abilityOwnerEntityId,
						IsTreasure: isTreasure,
					},
				),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/MercenariesAbilityActivatedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class MercenariesAbilityActivatedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public MercenariesAbilityActivatedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            Action action = null;
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(Action)
                && (action = node.Object as Action).Type == (int)BlockType.PLAY
                && GameState.CurrentEntities.ContainsKey(action.Entity)
                && GameState.CurrentEntities[action.Entity].GetZone() == (int)Zone.LETTUCE_ABILITY
                && GameState.CurrentEntities[action.Entity].GetCardType() == (int)CardType.LETTUCE_ABILITY;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var action = node.Object as Action;
            var entity = GameState.CurrentEntities[action.Entity];
            var controllerId = entity.GetEffectiveController();
            var cardId = entity.CardId;
            var abilityOwnerEntityId = entity.GetTag(GameTag.LETTUCE_ABILITY_OWNER);
            var isTreasure = entity.GetTag(GameTag.LETTUCE_IS_TREASURE_CARD) == 1;
            return new List<GameEventProvider> { GameEventProvider.Create(
                action.TimeStamp,
                "MERCENARIES_ABILITY_ACTIVATED",
                GameEvent.CreateProvider(
                    "MERCENARIES_ABILITY_ACTIVATED",
                    cardId,
                    controllerId,
                    entity.Id,
                    StateFacade,
                    //null,
                    new
                    {
                        AbilityOwnerEntityId = abilityOwnerEntityId,
                        IsTreasure = isTreasure,
                    }
                ),
                true,
                node) };
        }
    }
}
